lib/src/main/kotlin/com/swmansion/starknet/data/serializers/JsonRpcTransactionPayloadSerializer.kt

```diff
@@ -0,0 +1,53 @@
+package com.swmansion.starknet.data.serializers
+
+import com.swmansion.starknet.data.types.Felt
+import com.swmansion.starknet.data.types.transactions.*
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.Serializer
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.*
+
+@OptIn(ExperimentalSerializationApi::class)
+@Serializer(forClass = TransactionPayload::class)
+object JsonRpcTransactionPayloadSerializer : KSerializer<TransactionPayload> {
+
+    override val descriptor: SerialDescriptor
+        get() = TransactionPayload.serializer().descriptor
+
+    // TODO: Consider removing deserialize methods altogether
+    override fun deserialize(decoder: Decoder): TransactionPayload {
+        require(decoder is JsonDecoder)
+        val element = decoder.decodeJsonElement()
+
+        return when (element.jsonObject["type"]?.jsonPrimitive?.content) {
+            "INVOKE" -> decoder.json.decodeFromJsonElement(InvokeTransactionPayload.serializer(), element)
+            "DECLARE" -> deserializeDeclare(decoder, element)
+            "DEPLOY_ACCOUNT" -> decoder.json.decodeFromJsonElement(DeployAccountTransactionPayload.serializer(), element)
+            else -> throw IllegalArgumentException("Invalid transaction type '${element.jsonObject["type"]?.jsonPrimitive?.content}'")
+        }
+    }
+
+    private fun deserializeDeclare(decoder: JsonDecoder, element: JsonElement): DeclareTransactionPayload =
+            when (element.jsonObject["version"]?.jsonPrimitive?.content) {
+                Felt.ONE.hexString() -> decoder.json.decodeFromJsonElement(DeclareTransactionV1Payload.serializer(), element)
+                Felt(2).hexString() -> decoder.json.decodeFromJsonElement(DeclareTransactionV2Payload.serializer(), element)
+                else -> throw IllegalArgumentException("Invalid declare transaction version")
+            }
```

Comment:
Those were moved from the old `JsonRpcTransactionPayloadSerializer` but I'm not sure we ever need to deserialize `TransactionPayload` as this only something we use for building requests and never get from responses. 
There are no tests for that and I'm not sure we want to write these as this is something that is not going to be used.